File: spark_demo/src/java/com/sxt/java/core/AccumulatorOperator.java
package com.sxt.java.core;

import org.apache.spark.Accumulator;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.VoidFunction;
import org.apache.spark.rdd.RDD;

/**
 * 累加器在Driver端定义赋初始值和读取，在Executor端累加。
 * @author root
 *
 */
public class AccumulatorOperator {
	public static void main(String[] args) {

		SparkConf conf = new SparkConf();
		conf.setMaster("local").setAppName("accumulator");

		JavaSparkContext sc = new JavaSparkContext(conf);
		final Accumulator<Integer> accumulator = sc.accumulator(0);

		sc.textFile("data/word.txt",2).foreach(new VoidFunction<String>() {
			/**
			 * 
			 */
			private static final long serialVersionUID = 1L;

			@Override
			public void call(String t) throws Exception {
				accumulator.add(1);
//				System.out.println(accumulator.value());
				System.out.println(accumulator);
			}
		});


        // accumulator.value 写法只能在driver端，excutor端的task只能用accumulator 写法来查看数据

		System.out.println(accumulator.value());
		sc.stop();
		
	}
}

File: spark_demo/src/scala/com/sxt/scala/core/AccumulatorOperator.scala
package com.sxt.scala.core

import org.apache.spark.Accumulable
import org.apache.spark.SparkConf
import org.apache.spark.SparkContext

import org.apache.spark.deploy.master.Master
import org.apache.spark.deploy.worker.Worker

object AccumulatorOperator {

    def main(args: Array[String]): Unit = {
        val conf = new SparkConf()
        conf.setMaster("local").setAppName("accumulator")

        val sc = new SparkContext(conf)
        val accumulator = sc.accumulator(0)
        //java创建对象的几种方式： 1.new  2.clone 3.反射  4.反序列化
        var count = 0
        sc.textFile("data/word.txt",2).foreach { x => {
            count += 1
            println( " count : " + count)
        }}


        sc.textFile("data/word.txt",2).foreach { x => {
            accumulator.add(1)

        }}

        println("accumulator.value : " + accumulator.value)
        println("自定义 : " + count)
        sc.stop()
    }
}